kvbc/src/kvbc_app_filter/kvbc_app_filter.cpp

```diff
@@ -355,236 +354,197 @@ uint64_t KvbAppFilter::newestTagSpecificPublicEventGroupId() const {
+    return {global_eg_id, false, current_pvt_eg_id - 1, pub_eg_id};
+  }
+
+  ConcordAssert(external_event_group_id > current_ext_eg_id);
+  if (current_pvt_eg_id == private_end) {
+    auto num_pub_egs = current_ext_eg_id - current_pvt_eg_id;
+    auto pub_eg_id = num_pub_egs + (external_event_group_id - current_ext_eg_id);
+    std::tie(global_eg_id, std::ignore) =
+        getValueFromTagTable(kPublicEgId + kTagTableKeySeparator + concordUtils::toBigEndianStringBuffer(pub_eg_id));
+    return {global_eg_id, false, current_pvt_eg_id, pub_eg_id};
+  }
+  auto num_pub_egs = current_ext_eg_id - current_pvt_eg_id;
+  auto pub_eg_id = num_pub_egs + (external_event_group_id - current_ext_eg_id);
+  std::tie(global_eg_id, std::ignore) =
+      getValueFromTagTable(kPublicEgId + kTagTableKeySeparator + concordUtils::toBigEndianStringBuffer(pub_eg_id));
+  return {global_eg_id, false, current_pvt_eg_id, pub_eg_id};
 }
 
-void KvbAppFilter::readEventGroupRange(EventGroupId event_group_id_start,
-                                       spsc_queue<KvbFilteredEventGroupUpdate> &queue_out,
-                                       const std::atomic_bool &stop_execution) {
-  uint64_t public_start = getValueFromLatestTable(kPublicEgIdKeyOldest);
-  uint64_t private_start = getValueFromLatestTable(client_id_ + "_oldest");
-  uint64_t public_end = getValueFromLatestTable(kPublicEgIdKeyNewest);
-  uint64_t private_end = getValueFromLatestTable(client_id_ + "_newest");
-  LOG_DEBUG(logger_,
-            "Reading event group range, public_start: " << public_start << " private_start: " << private_start
-                                                        << " public_end: " << public_end
-                                                        << " private_end: " << private_end);
-  if (!public_start && !private_start) {
+void KvbAppFilter::readEventGroups(EventGroupId external_eg_id_start,
+                                   const std::function<bool(KvbFilteredEventGroupUpdate &&)> &process_update) {
+  uint64_t newest_public_eg_id = getNewestPublicEventGroupId();
+  uint64_t newest_private_eg_id = getValueFromLatestTable(client_id_ + "_newest");
+  uint64_t oldest_external_eg_id = oldestTagSpecificPublicEventGroupId();
+  uint64_t newest_external_eg_id = newestTagSpecificPublicEventGroupId();
+  if (not oldest_external_eg_id) {
     std::stringstream msg;
     msg << "Event groups do not exist for client: " << client_id_ << " yet.";
     LOG_ERROR(logger_, msg.str());
     throw std::runtime_error(msg.str());
   }
-  // update the offsets if we now have corresponding public/private event groups in storage
-  if (eg_data_state_->public_offset == 0) eg_data_state_->public_offset = public_start;
-  if (eg_data_state_->private_offset == 0) eg_data_state_->private_offset = private_start;
 
-  uint64_t event_group_id_end = private_end + public_end;
+  if (external_eg_id_start == 0) {
+    throw InvalidEventGroupId(external_eg_id_start);
+  }
```

Comment:
We could move this as the first check in the method as it depends on user input and we don't need to do any work if we are throwing. Another approach is to assert on 0 input.

Reply:
I will add an assert to the beginning. Thanks!